packages/react-router/modules/useRouter.js

```diff
@@ -0,0 +1,20 @@
+import React, { useContext } from "react";
+import invariant from "tiny-invariant";
+
+import RouterContext from "./RouterContext";
+import matchPath from "./matchPath";
+
+export default function useRouter(options = {}) {
+  invariant(
+    typeof useContext === "function",
+    "The useRouter hook requires React 16.7 or greater"
+  );
+
+  let context = useContext(RouterContext);
+  let location = options.location || context.location;
+  let match = options.path
+    ? matchPath(location.pathname, options)
```

Comment:
You could `useMemo` here like:

```js
let match = options.path
  ? useMemo(() => matchPath(location.pathname, options), [
      location.pathname,
      options.path,
      options.exact,
      options.strict
    ])
  : context.match;
```

So it won't loop through all routes at `matchPath` on every usage route remains unchanged after a render.

Reply:
I have a feeling that at the moment the `useMemo` came to the light, people are trying to memoize everything :) Bear in mind, that even such memoization has some overhead and without proper benchmarks, it's just a speculation what would be more performant.

Reply:
Oh and also conditional use of hook is a big NO-NO, so I really would do this.